brave/brave-ios
# Fix #5257: Wallet Ethereum Dapp Site Connections List + Dapp Settings

## Summary of Changes
- Add dapp preferences to wallet settings
- Add Manage Site Connections view for viewing all sites granted ethereum permissions

This pull request fixes #5257

## Submitter Checklist:


## Test Plan:
Note: At time of PR creation, wallet dapps are disabled with a compiler flag. Add `-DWALLET_DAPPS_ENABLED` to `base.xcconfig` to enable dapps support.

Manage Site Connections:
1. Visit some dapp sites like [Uniswap](https://app.uniswap.org/) / [Metamask Test Dapp](https://metamask.github.io/test-dapp/) and connect >1 accounts
5. Navigate to Wallet Settings and tap 'Manage site connections'
7. Tap 'Remove All', tap 'Remove' in the confirmation alert
8. Verify dapp sites had account connections removed, and were properly updated to show account connection removed (if tab was still open)
9. Repeat step 1 & 2
10. Open detail view for a single dapp site connection
11. Remove individual account from list
12. Verify dapp site(s) had the account connection removed, and were properly updated to show account connection removed (if tab was still open)

Default Wallet:
1. Visit Wallet Settings
2. Toggle 'Default Wallet' preference to 'None'
3. Visit any dapp site, verify wallet icon not shown in URL bar, no interaction with wallet is enabled via the dapp. [Metamask Test Dapp](https://metamask.github.io/test-dapp/) works well for testing this.
4. Visit Wallet Settings
5. Toggle 'Default Wallet' preference to 'Brave Wallet'
6. Visit any dapp site, verify wallet icon is shown in URL bar once account access or any other dapp request is received. [Uniswap](https://app.uniswap.org/) will prompt for account connection immediately (if no accounts connected), or can use [Metamask Test Dapp](https://metamask.github.io/test-dapp/) 

Allow Sites to Request Accounts:
1. Remove all ethereum connected accounts via 'Remove All' in 'Manage Site Connections'
2. Visit Wallet Settings & toggle 'Allow Sites to Request Accounts' to on.
3. Visit [Uniswap](https://app.uniswap.org/) and connect at least one account when prompted.
4. Visit Wallet Settings & toggle 'Allow Sites to Request Accounts' to off.
5. Visit [Metamask Test Dapp](https://metamask.github.io/test-dapp/) and tap 'Connect'. Verify you are not prompted to connect any accounts

Display Web3 Notifications:
1. Remove all ethereum connected accounts via 'Remove All' in 'Manage Site Connections'
2. Visit Wallet Settings & toggle 'Display Web3 Notifications' to off.
3. Visit [Uniswap](https://app.uniswap.org/), you should not see the purple wallet dapp notification / toast
4. Tap wallet icon in URL bar
5. Verify still prompted to connect account.
6. Visit Wallet Settings & toggle 'Display Web3 Notifications' to on. Remove Uniswap account connection if granted.
7. Visit [Uniswap](https://app.uniswap.org/), you should now see the purple wallet dapp notification / toast for Uniswap requesting wallet account permissions

## Screenshots:

![settings](https://user-images.githubusercontent.com/5314553/168342704-71a3a4ba-097b-4f66-a025-37ff9bb6c125.png) | ![manage site connections list](https://user-images.githubusercontent.com/5314553/168342719-4f7b342c-73ac-458d-b24b-773e0eae904b.png) | ![manage site connections](https://user-images.githubusercontent.com/5314553/168342938-c09711b5-d746-4736-87f9-0467e81cba1a.png)
--|--|--


## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `bug` / `enhancement`
